filters/filter.go

```diff
@@ -248,3 +292,243 @@ func (f *multiFilter) Matches(val string) bool {
 
 	return f.op == Conjunction
 }
+
+// chainFilter matches an input string against certain conditions
+// while returning the unmatched part of the input if there is a match
+type chainFilter interface {
+	fmt.Stringer
+
+	matches(val string) (bool, string)
+}
+
+// equalityChainFilter is a filter that performs equality string matches
+// from either the front or back of the string
+type equalityChainFilter struct {
+	pattern   string
+	backwards bool
+}
+
+func newEqualityChainFilter(pattern string, backwards bool) chainFilter {
+	return &equalityChainFilter{pattern: pattern, backwards: backwards}
+}
+
+func (f *equalityChainFilter) String() string {
+	return "Equals(\"" + f.pattern + "\")"
+}
+
+func (f *equalityChainFilter) matches(val string) (bool, string) {
+	if f.backwards && strings.HasSuffix(val, f.pattern) {
+		return true, val[:len(val)-len(f.pattern)]
+	}
+
+	if !f.backwards && strings.HasPrefix(val, f.pattern) {
+		return true, val[len(f.pattern):]
+	}
+
+	return false, ""
+}
+
+// singleAnyCharFilter is a filter that allows any one char
+type singleAnyCharFilter struct {
+	backwards bool
+}
+
+func newSingleAnyCharFilter(backwards bool) chainFilter {
+	if backwards {
+		return singleAnyCharFilterBackwards
+	}
+
+	return singleAnyCharFilterForwards
+}
+
+func (f *singleAnyCharFilter) String() string { return anyCharStr }
+
+func (f *singleAnyCharFilter) matches(val string) (bool, string) {
+	if len(val) == 0 {
+		return false, ""
+	}
+
+	if f.backwards {
+		return true, val[:len(val)-1]
+	}
+
+	return true, val[1:]
+}
+
+// newSingleRangeFilter creates a filter that performs range matching
+// on a single char
+func newSingleRangeFilter(pattern string, backwards bool) (chainFilter, error) {
+	if len(pattern) == 0 {
+		return nil, errInvalidFilterPattern
+	}
+
+	negate := false
+	if pattern[0] == negationChar {
+		negate = true
+		pattern = pattern[1:]
+	}
+
+	if len(pattern) == 3 && pattern[1] == rangeChar {
```

Comment:
I didn't see it in the standard glob definition, but I'll add it in the next diff - good example of a use case there I guess.